add StatClient

ReViZe96 opened the pull request:
Добавлен клиент статистики

ReViZe96 committed:
add StatClient

ReViZe96 committed:
checkstyle on static field

ReViZe96 committed:
checkstyle on static field 2

iolinpic commented on stat/stat-client/pom.xml:
```diff
@@ -44,6 +44,12 @@
+            <groupId>ru.practicum</groupId>
+            <artifactId>stat-dto</artifactId>
+            <version>0.0.1-SNAPSHOT</version>
+            <scope>compile</scope>
```
compile совсем не обязателен для модулей которых подключаем локально

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/StatisticClient.java:
```diff
@@ -4,10 +4,10 @@
 
 @SpringBootApplication
-public class StatClient {
+public class StatisticClient {
```
Зачем? (не то чтобы надо все вернуть, но просто какая мотивация?

ReViZe96 replied:
Класс реализации клиента теперь называется StatClient. Чтобы избежать дублирования названий классов в модуле stat-client, main-класс пришлось переименовать в StatisticClient

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/client/StatClient.java:
```diff
@@ -0,0 +1,45 @@
+
+public class StatClient {
+
+    private final String baseUrl = "http://localhost:9090";
```
это точно надо выносить в настройки - во первых при запуске в докере путь будет другой (в нашем случае http://stats-server:9090) а для запуска снаружи в самый раз - сюда бы как раз подошло задание через параметры контроллера например

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/client/StatClient.java:
```diff
@@ -0,0 +1,45 @@
+                .body(new ParameterizedTypeReference<>() {
+                });
+    }
+
```
в целом по запросам все вроде понятно - но что будет если в ответ будет какаянить ошибка? - по идее их надо тоже ловить - по крайней мере те что указаны в спеке.

iolinpic commented on stat/stat-server/src/main/java/ru/practicum/StatisticServer.java:
```diff
@@ -4,10 +4,10 @@
 
 @SpringBootApplication
-public class StatServer {
+public class StatisticServer {
```
аналогичный вопрос - зачем? вдвойне вопрос потому что не относится к реализуемой фиче.

ReViZe96 replied:
Поскольку теперь в модуле stat-client main-класс называется StatisticClient, main-класс модуля stat-server переименовал по аналогии (у main-классов статистики - префикс Statistic)

iolinpic reviewed (changes requested):
точно нужно чтото сделать с хардкодом базового пути и возможными ошибками

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/client/StatClient.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void saveHit(EndpointHitDto hitDto) {
+        restClient.post()
+                .uri(baseUrl)
```
тут пропущен baseUrl+"/hit"

ReViZe96 committed:
1. Добавлен эндпоинт /hit
2. Параметризован базовый хост
3. Обработка возможных ошибок: оборачиваем в ClientException и пробрасываем

ReViZe96 committed:
add ClientException

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/exception/ClientException.java:
```diff
@@ -0,0 +1,15 @@
+        this.status = status;
+        this.description = description;
+    }
+
```
вообще в таких случаях рекомендуется использовать super(String message) хотябы - чтобы во всех стандартных способах это отобразилось ( и не пилить лишнего руками самим)

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/client/StatClient.java:
```diff
@@ -0,0 +1,92 @@
+                            response.getStatusCode().value(),
+                            response.getBody().toString()
+                    );
+                })
```
нет перехвата ошибки, когда сервер вообще ничего не вернул ( например если мы не правильно урл записали)

iolinpic commented on stat/stat-client/src/main/java/ru/practicum/client/StatClient.java:
```diff
@@ -0,0 +1,92 @@
+        return restClient.get()
+                .uri(uriWithParams).retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, (request, response) -> {
+                    throw new ClientException(
```
учитывая что 4 раза запускается абсолютно идентичная функция, я бы ее сохранил отдельно как функцию и просото вызывад

iolinpic reviewed (approved):
В целом я бы доработал (но можно и ревьюера порадовать)
